backend/globaleaks/utils/import_export.py

```diff
@@ -0,0 +1,419 @@
+    '''
+
+    if tid is None:
+        tenant_data['tenant'].id = None
+        tenant_obj = session.merge(tenant_data['tenant'])
+        session.flush()
+        tid = tenant_obj.id
+    else:
+        tenant_data['tenant'].id = tid
+        session.merge(tenant_data['tenant'])
+
+    # Correct the TID in all rows
+    for datatype, rowset in tenant_data.items():
+        if datatype is 'tenant':
+            continue
+
+        for row in rowset:
+            if hasattr(row, 'tid'):
+                row.tid = tid
+            if hasattr(row, 'tenant_id'):
+                row.tenant_id = tid
+
+    # Replay the tenant data
+    for element in IMPORT_ORDER:
+        if element == 'fieldanswer':
+            # fieldanswer needs to deNULL the above
+            for fieldanswer in tenant_data['fieldanswer']:
+                nulled_field = session.query(models.FieldAnswer).filter_by(id=fieldanswer.id).first()
+                nulled_field.fieldanswergroup_id = fieldanswer.fieldanswergroup_id
+        else:
+            for row in tenant_data[element]:
+                session.merge(row)
+
+    session.commit()
+    session.close()
+
+def create_export_tarball(session, tid):
+    '''Creates an export tarball, either as a file on disk, or in memory as a variable'''
+
+    # Collect tenant data
+    tenant_data = collect_all_tenant_data(session, tid)
+
+    try:
+        dirpath = tempfile.mkdtemp()
+
+        # Create the export database
+        write_tenant_to_fresh_db(tenant_data, dirpath + "/globaleaks.db")
+
+        # Write out a export format version marker
+        with open(dirpath + "/EXPORT_FORMAT", 'w') as f:
+            f.write(str(1))
+
+        output_file = io.BytesIO()
+
+        with tarfile.open(fileobj=output_file, mode='w:gz') as export_tarball:
+            export_tarball.add(dirpath + "/EXPORT_FORMAT", arcname="EXPORT_FORMAT")
+            export_tarball.add(dirpath + "/globaleaks.db", arcname="globaleaks.db")
+
+            def process_files_for_tarball(fileset):
+                for file_obj in fileset:
+                    tarball_file = "attachments/"+file_obj.filename
+                    file_to_read = os.path.join(Settings.attachments_path, file_obj.filename)
+                    try:
+                        export_tarball.add(file_to_read, arcname=tarball_file)
+                    except FileNotFoundError:
+                        # Reference files might not exist if the system
+                        # successfully encrypted them for all receivers
+
+                        if hasattr(file_obj, 'status'):
+                            if file_obj.status == u'reference':
+                                print("Reference file " + file_to_read + " not found, skipping.")
+
+            process_files_for_tarball(tenant_data['receiverfile'])
+            process_files_for_tarball(tenant_data['whistleblowerfile'])
+
+    finally:
+        shutil.rmtree(dirpath)
+
+    return output_file.getvalue()
```

Comment:
@NCommander:
This function is correct but the specific handling of this getvalue and of the Bytes buffer make it write it all in RAM.
Use a real file in place of a RAM buffer and return the file handle.